### PR TITLE
Bugfix - exchangeCapabilites based on ChainSpec

### PR DESCRIPTION
**Pull request by MarekM25**
## Changes

Correct fix to @flcl42 workaround hotfix https://github.com/NethermindEth/nethermind/pull/5787 

If Canun is not set in chainspec we won't send V3 methods
If Shanghai is not set in chainspec we won't send V2 methods etc

Let's be careful about gnosis because it will mean that we will use V1 for gnosis.

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

Before merging requires testing with gnosis nodes and one mainnet. I will do it after the reviews

### CODE REVIEW

File: src/Nethermind/Nethermind.Core/Specs/ISpecProvider.cs [hunk 1]
```diff
@@ -75,7 +75,10 @@ public interface ISpecProvider
         /// <summary>
         /// Resolves a spec for all planned forks applied.
         /// </summary>
-        /// <returns>A spec for all planned forks applied</returns>
-        IReleaseSpec GetFinalSpec() => GetSpec(long.MaxValue, ulong.MaxValue);
+        /// <returns>A spec for all planned forks applied</
+        /// The default value is long.MaxValue for block numbers and ulong.MaxValue for timestamps
+        /// for every new not yet scheduled EIP. Because of that we can't use long.MaxValue and
+        /// ulong.MaxValue for GetFinalSpec that is why we have long.MaxValue-1, ulong.MaxValue-1
+        IReleaseSpec GetFinalSpec() => GetSpec(long.MaxValue-1, ulong.MaxValue-1);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Specs/ISpecProvider.cs:82]: Use <remarks> ?

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.Setup.cs [hunk 2]
```diff
@@ -52,9 +52,8 @@ public Task Setup()
         IPayloadPreparationService? mockedPayloadService = null, ILogManager? logManager = null) =>
         new(mergeConfig, mockedPayloadService, logManager);
 
-    protected async Task<MergeTestBlockchain> CreateShanghaiBlockChain(IMergeConfig? mergeConfig = null,
-        IPayloadPreparationService? mockedPayloadService = null)
-        => await CreateBlockChain(mergeConfig, mockedPayloadService, Shanghai.Instance);
+    protected async Task<MergeTestBlockchain> CreateBlockChain(IReleaseSpec? releaseSpec, IMergeConfig? mergeConfig = null)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.Setup.cs:55]: `Blockchain`

File: src/Nethermind/Nethermind.Merge.Plugin/Handlers/EngineRpcCapabilitiesProvider.cs [hunk 3]
```diff
@@ -11,38 +11,38 @@ namespace Nethermind.HealthChecks;
 
 public class EngineRpcCapabilitiesProvider : IRpcCapabilitiesProvider
 {
-    private static readonly ConcurrentDictionary<string, bool> _capabilities = new();
+    private readonly ConcurrentDictionary<string, (bool Enabled, bool WarnIfMissing)> _capabilities = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Handlers/EngineRpcCapabilitiesProvider.cs:14]: I am thinking enum flags might be better than Tuple?

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Core/Specs/ISpecProvider.cs (hunk 1 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Core/Specs/ISpecProvider.cs:82]: hmmm, what do you mean?

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.Setup.cs (hunk 2 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.Setup.cs]: done

File: src/Nethermind/Nethermind.Core/Specs/ISpecProvider.cs [hunk 4]
```diff
@@ -75,7 +75,10 @@ public interface ISpecProvider
         /// <summary>
         /// Resolves a spec for all planned forks applied.
         /// </summary>
-        /// <returns>A spec for all planned forks applied</returns>
-        IReleaseSpec GetFinalSpec() => GetSpec(long.MaxValue, ulong.MaxValue);
+        /// <returns>A spec for all planned forks applied</
+        /// <remarks> The default value is long.MaxValue for block numbers and ulong.MaxValue for timestamps
+        /// for every new not yet scheduled EIP. Because of that we can't use long.MaxValue and
+        /// ulong.MaxValue for GetFinalSpec that is why we have long.MaxValue-1, ulong.MaxValue-1 </remarks>
+        IReleaseSpec GetFinalSpec() => GetSpec(long.MaxValue-1, ulong.MaxValue-1);
```
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.Core/Specs/ISpecProvider.cs:82]: You need spaces in `long.MaxValue - 1` to pass whitespaces check

**Comment by MarekM25** [comment]
> What's the advantage of this approach over the one of #5787? In other words, why listing a capability in the dictionary with an extra flag is better than not listing it at all?

You want to have the same code for 4844 devnet and for mainnet and at the same time you have to serve {V1, V2} for mainnet and {V1,V2,V3} for 4844 devnet. @flcl42 PR will work only for mainnet

**Pull request merged**